### PR TITLE
Tests autodiscovery replacing hardcoded file list

### PR DESCRIPTION
**Pull request by evemorgen**
### Description

This PR hopefully resolves https://github.com/python/mypy/issues/8650.

The main change is the `find_test_files ` helper function that looks for files matching the pattern in the unit tests directory. It has been used in the following test files for autodiscovery:
-  `testcheck.py`
-  `testdeps.py`
- ` testfinegrained.py`
- `testparse.py`
- `testsemanal.py`

I've also updated the readme. 

## Test Plan

This is a question for maintainers - does this helper function needs unit tests and if so, where should they live? I haven't found `testhelpers.py` 🙂

### CODE REVIEW

File: mypy/test/helpers.py [hunk 1]
```diff
@@ -494,3 +495,11 @@ def normalize_file_output(content: List[str], current_abs_path: str) -> List[str
     result = [re.sub(r'\b' + re.escape(base_version) + r'\b', '$VERSION', x) for x in result]
     result = [timestamp_regex.sub('$TIMESTAMP', x) for x in result]
     return result
+
+
+def find_test_files(pattern: str, exclude: Optional[List[str]] = None) -> List[str]:
+    return [
+        str(x.name)
```
Comment (odesenfans) [inline mypy/test/helpers.py:502]: `x.name` is already a string, you can remove the conversion.

Comment (ilevkivskyi) [reply mypy/test/helpers.py:502]: Also maybe use a bit more descriptive name than `x` (`path` maybe?)

Comment (ilevkivskyi) [reply mypy/test/helpers.py:502]: @evemorgen Could you please implement these two suggestions? ^^ I will then merge this.

Comment (evemorgen) [reply mypy/test/helpers.py]: @ilevkivskyi Done in [7974ae6](https://github.com/python/mypy/pull/13143/commits/7974ae67cf3ca261d585c9aba5032b620274dadc) 🙂

**Comment by evemorgen** [comment]
@ilevkivskyi I think we're all good to merge it now 🙂

**Review by ilevkivskyi — APPROVED** [review]

**Comment by hauntsaninja** [comment]
Since https://github.com/python/mypy/pull/13138 silently broke half our tests, I thought I'd double check this as well.

Tests for Python 3.10 on 6f93fc16e (with this PR)
```
========== 10142 passed, 375 skipped, 9 xfailed in 1112.76s (0:18:32) ==========
```
Tests for Python 3.10 on b6d525d3 (without this PR)
```
========== 10144 passed, 375 skipped, 9 xfailed in 1034.39s (0:17:14) ==========
```
Any idea what 2 tests we lost?

**Comment by hauntsaninja** [comment]
cc @evemorgen

**Comment by evemorgen** [comment]
I'll look into that later today

**Comment by evemorgen** [comment]
@hauntsaninja Yeah, that's right, it looks the like two tests have been lost for any reason. I'll post a PR with a fix shortly.
```diff
> diff test-310-pr.log test-310-prev.log
5272a5273,5274
>         <DataDrivenTestCase testLiteralSemanalBasicAssignment>
>         <DataDrivenTestCase testLiteralSemanalInFunction>
10311a10314
10879c10882
< ======================== 10537 tests collected in 0.46s ========================
---
> ======================== 10539 tests collected in 0.42s ========================
```

**Comment by evemorgen** [comment]
@hauntsaninja - https://github.com/python/mypy/pull/13175 this PR reintroduces missing tests

**Pull request merged**